Sources/SwiftParser/Declarations.swift

```diff
@@ -1120,18 +1120,45 @@ extension Parser {
     _ handle: RecoveryConsumptionHandle
   ) -> RawDeinitializerDeclSyntax {
     let (unexpectedBeforeDeinitKeyword, deinitKeyword) = self.eat(handle)
+
     var unexpectedNameAndSignature: [RawSyntax?] = []
-    unexpectedNameAndSignature.append(self.consume(if: TokenSpec(.identifier, allowAtStartOfLine: false)).map(RawSyntax.init))
-    if self.at(.leftParen) && !self.currentToken.isAtStartOfLine {
-      unexpectedNameAndSignature.append(RawSyntax(parseFunctionSignature()))
+
+    // async is a contextual keyword
+    // must be parsed before attempting to parse identifier
+    var effectSpecifiers = parseDeinitEffectSpecifiers()
+
+    if effectSpecifiers == nil {
+      if let identifier = self.consume(if: TokenSpec(.identifier, allowAtStartOfLine: false)).map(RawSyntax.init) {
+        unexpectedNameAndSignature.append(identifier)
+      }
+      effectSpecifiers = parseDeinitEffectSpecifiers()
+    }
+    if effectSpecifiers == nil && self.at(TokenSpec(.leftParen, allowAtStartOfLine: false)) {
+      let input = parseParameterClause(RawParameterClauseSyntax.self) { parser in
+        parser.parseFunctionParameter()
+      }
+      unexpectedNameAndSignature.append(RawSyntax(input))
+
+      effectSpecifiers = parseDeinitEffectSpecifiers()
     }
+
+    var unexpectedAfterAsync: [RawSyntax?] = []
+    /// Only allow recovery to the arrow with exprKeyword precedence so we only
+    /// skip over misplaced identifiers and don't e.g. recover to an arrow in a 'where' clause.
+    if self.at(.arrow) || self.canRecoverTo(TokenSpec(.arrow, recoveryPrecedence: .exprKeyword)) != nil {
```

Comment:
`self.at(.arrow)` implies `self.canRecoverTo(TokenSpec(.arrow, recoveryPrecedence: .exprKeyword)) != nil` so I think you can just remove `self.at(.arrow)`.

Reply:
I've copied this from `parseFunctionSignature()`. Should it be updated there as well? Could it be that this pre-check is a performance optimization? Any other reasons it might be needed there?

Reply:
If you could also update the call in `parseFunctionSignature`, that would be great. I don’t see how `self.at(.arrow)` could help performance in any significant way.

Reply:
I think you only updated the existing call and forgot to update this one 😉

Reply:
🤦‍♂️